Allow to fetch multiple config parameters as a Map

essobedo opened the pull request:
fixes #578 

## Motivation

When we need a dynamic configuration, it can be very helpful to use a map. For example, let's say that I want to keep in my configuration the custom reason phrases to return to the enduser according to the http status code.

I would like to have:

The expected code

```
@ConfigProperty(name = "reasons")
Map<Integer, String> reasonPhrases; // or eventually Map<String, String>
```

The expected configuration
```
reasons.200=My custom reason phrase for OK
reasons.201=My custom reason phrase for Created
...
```

## Modifications

* Adds a generic producer for Map
* Adds a MapConverter to convert the content of type `<key1>=<value1>;<key2>=<value2>...` into a `Map`.
* Adds map support to the `ConfigMappingGenerator` 
* Adds related unit tests
* Adds new methods `getValues(String name, Class<K> kClass, Class<V> vClass)` and `getOptionalValues(String name, Class<K> kClass, Class<V> vClass)` to support properly the map from the class `SmallRyeConfig`

essobedo commented:
Hi @radcortez it is the PR related https://github.com/eclipse/microprofile-config/issues/562#issuecomment-841279200
It is very basic because there are many things unspecified (like for example how to manage default values in case of `@ConfigProperty`) but at least it covers simple uses cases. WDYT of this PR?

radcortez commented on cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:
```diff
@@ -107,6 +111,36 @@ private ConfigProducerUtil() {
+            ParameterizedType paramType = (ParameterizedType) type;
+            Type[] typeArgs = paramType.getActualTypeArguments();
+            for (String pName : config.getPropertyNames()) {
+                if (pName.startsWith(prefix)) {
```
What if you have multiple subnamespaces? For instance:

```
my.map.key=key
my.map.value=value
my.map.sub.something=1234
```

essobedo replied:
@radcortez You mean that in that case you would expect only the keys `key` and `value` in the map with `@ConfigProperty(name = "my.map")`?

radcortez replied:
I guess this was not a good example :) 

My point was about keys after the map that have more that one level? Should the key be included as `x.y`? Should it be ignored because it is composed?

essobedo replied:
I understand your point, for me since there many unspecified  aspects like this one, I wanted to keep it simple by making a "flat map". 
So in your example  If I have the following code:

```
@ConfigProperty(name = "my.map")
Map<String, String> map;
```

I expect to have key/key, value/value and sub.something/1234 in my map as pairs of String. 

Subnamespaces could make sense if we wanted to support complex mapping like:

```
my.map.c1.name=John
my.map.c1.age=33
```

The target class:
```
class Customer {
   String name;
   int age;
}
```

The mapping definition

```
@ConfigProperty(name = "my.map")
Map<String, Customer> map;
```

I could then expect to have one entry inside my map whose key is `c1` but I think that we would be too far from the current logic of the original spec where the value of all fields annotated with `@ConfigProperty` should be the value of the property converted into the target type.

Don't you agree?

radcortez replied:
I don't think we should support complex mapping (well we could, since this is already supported via @ConfigMapping, but I believe it would require a lot of extra work)

What I'm trying to define here is what should be the behaviour in a case we find a subnamespace when injecting a map. I would say that we should ignore these properties.

essobedo replied:
@radcortez Fixed

radcortez commented on cdi/src/test/java/io/smallrye/config/inject/ConfigInjectionTest.java:
```diff
@@ -80,6 +83,9 @@ void converters() {
     static class ConfigBean {
+        @Inject
+        @ConfigProperty(name = "reasons")
+        Map<Integer, String> reasons;
```
I think we must add some more tests. Add more Map combinations, use a custom converter, etc.

essobedo replied:
@radcortez I added some tests, let me know if it is not enough. Please note that it brings support of Map of List or Map of Set but only with `@ConfigProperty` but not in case of `@ConfigProperties`, if you expect the same support with `@ConfigProperties`, I think that it should be done in another ticket.

essobedo replied:
@radcortez done

radcortez reviewed (changes requested):
Thanks for the PR!

> how to manage default values in case of `@ConfigProperty`, 
We could support something like `key=value` and parse the default value if the injection is a map. Alternatively, we could just disallow it all together and throw a validation exception.

> how to deal with `config.getValue` in case of map since we have no way to get the Parameterized types) 
We would need to support something similar to a `GenericType` like JAX-RS to carry type information, but for now we could just assume that the programmatic API for Map returns a `Map<String,String>`.

essobedo commented:
> how to manage default values in case of `@ConfigProperty`,
> We could support something like `key=value` and parse the default value if the injection is a map. Alternatively, we could just disallow it all together and throw a validation exception.

I just added the support of default value with the next format `<key1>=<value1>;<key2>=<value2>...`, knowing that we can use a backslash to escape the characters `=` and `;` if needed

> how to deal with `config.getValue` in case of map since we have no way to get the Parameterized types)
> We would need to support something similar to a `GenericType` like JAX-RS to carry type information, but for now we could just assume that the programmatic API for Map returns a `Map<String,String>`.

I added a new BuiltIn Converter for the Map to be able to support `config.getValue`

essobedo commented:
@radcortez Hi, all your remarks have been addressed, this PR is ready for review now. Thank you in advance 🙏

essobedo commented:
> I added a new BuiltIn Converter for the Map to be able to support `config.getValue`

Finally I rather added new dedicated methods as extensions for the map support such that I don't need the Map converter to be part of the Builtin converters

radcortez commented:
> I expected to have it included into Quarkus 2.0, I guess it is too late now 😢

I wouldn't be too worried about it. Quarkus has a very high release cadence, so it we miss the boat now, we will get the next one soon :)

Anyway, we still need to test this in Quarkus. Did you try it? In the meantime I'll have review the PR.

radcortez commented on implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:
```diff
@@ -191,13 +191,44 @@
+     * @throws NoSuchElementException if no direct sub properties could be found.
+     */
+    @Experimental("Extension to retrieve mandatory sub properties as a Map")
+    public <K, V> Map<K, V> getMapValue(String name, Class<K> kClass, Class<V> vClass) {
```
Wondering if we should just name this `getValue`. It shouldn't clash with any of the other methods and it will be consistent with `getValue` and `getValues`. What do you think?

essobedo replied:
If so I would rather use `getValues` since in a `Map` you have more than one value, don't you agree?

radcortez commented on implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:
```diff
@@ -294,11 +325,68 @@ public String getRawValue(String name) {
+     * @return the resolved property value as an instance of the requested Map (not {@code null})
+     * @throws IllegalArgumentException if a key or a value cannot be converted to the specified types
+     */
+    @Experimental("Extension to retrieve non mandatory sub properties as a Map")
```
Same here with the naming.

radcortez commented on implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:
```diff
@@ -294,11 +325,68 @@ public String getRawValue(String name) {
+     * @return a String containing the matching properties that have been formatted if any, {@code null}
+     *         otherwise.
+     */
+    public String getMapRawValue(String name) {
```
Wondering if there is any point in flattening the map values to a single String, just so we can use the Converter. Maybe we could just create and populate the map directly?

essobedo replied:
I know but the api is String based, I mean the converters expect String, the ConfigValue expect String, the default value is expected as a String... If really needed I can check how it can be improved in a different commit

radcortez replied:
It can be done like the indexed property support. Lets leave it for now and come back to it later if it causes any issues.

essobedo replied:
@radcortez Done in a different commit to revert it easily if needed. Please tell me if it is more or less what you had in mind, meanwhile I will test it with Quarkus then I will add some doc about it

radcortez reviewed (commented):
I think this is going into the right direction. Left a few comments for us to discuss :) Thanks!

essobedo commented on cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:
```diff
@@ -208,6 +222,28 @@ private static String resolveDefault(String rawValue, String defaultValue) {
+     * @return {@code true} if the given type is a type of Map or is a Supplier or Optional of Map,
+     *         {@code false} otherwise.
+     */
+    private static boolean isMapExpected(final Type type) {
```
You mean `hasCollection` right?

radcortez commented:
@essobedo Thank you! I think we are almost done. Just one last detail: can we document the new feature here: https://github.com/smallrye/smallrye-config/blob/fecdcf541a0ad979cc3f83f8e5f4829da61e5be8/doc/modules/ROOT/pages/config/config.adoc. You can follow the indexed properties example.

radcortez commented:
Also, I recommend to make a build and test it with Quarkus master. Did you try it with Quarkus as is?

essobedo commented:
> Also, I recommend to make a build and test it with Quarkus master. Did you try it with Quarkus as is?

Not yet but I will test it today and let you know the results.

essobedo commented on implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:
```diff
@@ -191,13 +191,75 @@
+     * @return the resolved property value as an instance of the requested Map or {@code null} if it could not be found.
+     * @throws IllegalArgumentException if a key or a value cannot be converted to the specified types
+     */
+    public <K, V> Map<K, V> valuesAsMap(String name, Converter<K> keyConverter, Converter<V> valueConverter) {
```
I could not use getValues as name here because it is in conflict with `getValues(String name, Converter<T> converter, IntFunction<C> collectionFactory)`

radcortez replied:
Ok, let's just use `getValuesAsMap`. So we can keep the `get` consistency.

essobedo commented:
@radcortez just to let you know that I had to make couple of additional fixes to make it work with Quarkus, knowing that even on Quarkus side there are some little changes to make. I have not yet finished but I could make injection with `@ConfigProperty` work but I have one last bug to fix on Quarkus side with `@ConfigProperties`. Indeed I get the following error at runtime only, it should not be too hard to fix:

```
2021-05-28 20:47:41,177 ERROR [io.qua.run.Application] (main) Failed to start application (with profile worker): java.lang.ClassNotFoundException: org.objectweb.asm.ClassVisitor
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:104)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:60)
        at io.smallrye.config.ConfigMappingClass.<init>(ConfigMappingClass.java:35)
        at io.smallrye.config.ConfigMappingClass.createConfigurationClass(ConfigMappingClass.java:27)
        at io.smallrye.config.ConfigMappingClass.access$000(ConfigMappingClass.java:7)
        at io.smallrye.config.ConfigMappingClass$1.computeValue(ConfigMappingClass.java:11)
        at io.smallrye.config.ConfigMappingClass$1.computeValue(ConfigMappingClass.java:8)
        at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:228)
        at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:210)
        at java.base/java.lang.ClassValue.get(ClassValue.java:116)
        at io.smallrye.config.ConfigMappingClass.getConfigurationClass(ConfigMappingClass.java:17)
        at io.smallrye.config.ConfigMappingLoader.getConfigMappingClass(ConfigMappingLoader.java:42)
        at io.smallrye.config.ConfigMappingProvider$Builder.addRoot(ConfigMappingProvider.java:887)
        at io.smallrye.config.ConfigMappings.mapConfiguration(ConfigMappings.java:52)
        at io.smallrye.config.ConfigMappings.registerConfigProperties(ConfigMappings.java:43)
        at io.quarkus.arc.runtime.ConfigRecorder.registerConfigProperties(ConfigRecorder.java:60)
        at io.quarkus.deployment.steps.ConfigBuildStep$registerConfigClasses-2116437784.deploy_0(ConfigBuildStep$registerConfigClasses-2116437784.zig:159)
        at io.quarkus.deployment.steps.ConfigBuildStep$registerConfigClasses-2116437784.deploy(ConfigBuildStep$registerConfigClasses-2116437784.zig:40)
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:546)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:101)
```

essobedo commented:
@radcortez After a deeper investigation, it appears that I need to move the method `generateInterfaceName` out of `ConfigMappingGenerator` to avoid loading `ClassVisitor` at runtime for nothing since the method doesn't use it and the byte code has been generated at build time already

essobedo commented on implementation/src/main/java/io/smallrye/config/ConfigMappingGenerator.java:
```diff
@@ -352,20 +352,6 @@
         return writer.toByteArray();
     }
 
-    static String generateInterfaceName(final Class<?> classType) {
```
Need to move it to prevent getting `ClassNotFoundException` at runtime because `asm` is missing which is wanted as not used and not expected in native mode

radcortez replied:
Interesting. I wonder why we didn't pick this up before.

essobedo commented:
So I tested it successfully on Quarkus (modulo a small patch to support map) with the next use case:

The main configuration class:
```
@Singleton
public class GreetingConfiguration {

    @ConfigProperty(name = "root.numbers")
    public Map<Integer, String> numbers;

    @ConfigProperty(name = "root.numbers")
    public Optional<Map<Integer, String>> oNumbers;

    @ConfigProperty(name = "root.numbers")
    public Supplier<Map<Integer, String>> sNumbers;

    @ConfigProperties
    public RootConfig root;
}
```

The class `RootConfig`:
```
@ConfigProperties(prefix = "root")
public class RootConfig {

    public Map<Integer, String> numbers;
}
```

Let me know if it is good enough for you or you have another uses in mind to test, meanwhile I start working on the doc.

essobedo commented:
FYI here is my test branch for Quarkus https://github.com/quarkusio/quarkus/compare/main...essobedo:add-map-support?expand=1

essobedo commented:
@radcortez I added the doc, so now all your remarks have been addressed

radcortez commented on cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:
```diff
@@ -128,20 +142,20 @@ static String getRawValue(String name, Config config) {
     }
 
     private static String resolveValue(String name, String defaultValue, Config config) {
```
This seems to be used only by the indexed case and called always with null as a default. We should inline any code and remove the method.

radcortez commented on implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:
```diff
@@ -191,13 +191,75 @@
+     *         empty string or the converter returns {@code null}
+     */
+    @SuppressWarnings("unchecked")
+    private <T> T getValue(String name, ConfigValue configValue, Converter<T> converter) {
```
I think we don't need to split the method anymore, since it only has one usage now.

radcortez reviewed (changes requested):
@essobedo Thank you for your patience. Very good job. Just a few more minor things (leftovers from the other proposals and we should be good to go).

essobedo commented:
@radcortez Done, please check again. Do you want me to squash my commits?

radcortez commented:
> @radcortez Done, please check again. Do you want me to squash my commits?

Yes, please. Thank you :)

essobedo commented:
> > @radcortez Done, please check again. Do you want me to squash my commits?
> 
> Yes, please. Thank you :)

Done

radcortez commented:
@essobedo Thank you very much for all the work. I hope you can continue to contribute :)

Can you please open a Draft PR for Quarkus with the changes required? It it fine to use the SNAPSHOT (the CI won't build it as draft). We can update the version once we release, so the PR is for us to not forget the changes.

I think we may need some other fixes for Quarkus in other areas, so we need these in before doing the final release.

essobedo commented:
> @essobedo Thank you very much for all the work. I hope you can continue to contribute :)

Your welcome, it is always with great pleasure that I contribute to open source projects. If you would like me to help you (if I can) on another ticket, please let me know 😄 

> Can you please open a Draft PR for Quarkus with the changes required? It it fine to use the SNAPSHOT (the CI won't build it as draft). We can update the version once we release, so the PR is for us to not forget the changes.
> 
> I think we may need some other fixes for Quarkus in other areas, so we need these in before doing the final release.

Here it is https://github.com/quarkusio/quarkus/pull/17579

rmanibus commented:
Thanks a lot @essobedo !
